### ISSUE TITLE
CT Trail visibility failure

### ISSUE THREAD

**Issue by KatieWoe** [phetsims/center-and-variability#191]
```
center-and-variability : fuzz : unbuilt : assertSlow
http://127.0.0.1/continuous-testing/ct-snapshots/1683626850194/center-and-variability/center-and-variability_en.html?continuousTest=%7B%22test%22%3A%5B%22center-and-variability%22%2C%22fuzz%22%2C%22unbuilt%22%2C%22assertSlow%22%5D%2C%22snapshotName%22%3A%22snapshot-1683626850194%22%2C%22timestamp%22%3A1683628277428%7D&brand=phet&eall&fuzz
Query: brand=phet&eall&fuzz
Uncaught Error: Assertion failed: Trail visibility failure
Error: Assertion failed: Trail visibility failure
at window.assertions.assertFunction (http://127.0.0.1/continuous-testing/ct-snapshots/1683626850194/assert/js/assert.js:28:13)
at assertSlow (Instance.js:1886:6)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
id: Sparky Puppeteer
Snapshot from 5/9/2023, 4:07:30 AM

----------------------------------

center-and-variability : fuzz : unbuilt : assertSlow
http://127.0.0.1/continuous-testing/ct-snapshots/1683626850194/center-and-variability/center-and-variability_en.html?continuousTest=%7B%22test%22%3A%5B%22center-and-variability%22%2C%22fuzz%22%2C%22unbuilt%22%2C%22assertSlow%22%5D%2C%22snapshotName%22%3A%22snapshot-1683626850194%22%2C%22timestamp%22%3A1683631787584%7D&brand=phet&eall&fuzz
Query: brand=phet&eall&fuzz
Uncaught Error: Assertion failed: Trail visibility failure
Error: Assertion failed: Trail visibility failure
at window.assertions.assertFunction (http://127.0.0.1/continuous-testing/ct-snapshots/1683626850194/assert/js/assert.js:28:13)
at assertSlow (Instance.js:1886:6)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
id: Sparky Puppeteer
Snapshot from 5/9/2023, 4:07:30 AM

----------------------------------

center-and-variability : fuzz : unbuilt : assertSlow
http://127.0.0.1/continuous-testing/ct-snapshots/1683626850194/center-and-variability/center-and-variability_en.html?continuousTest=%7B%22test%22%3A%5B%22center-and-variability%22%2C%22fuzz%22%2C%22unbuilt%22%2C%22assertSlow%22%5D%2C%22snapshotName%22%3A%22snapshot-1683626850194%22%2C%22timestamp%22%3A1683633930634%7D&brand=phet&eall&fuzz
Query: brand=phet&eall&fuzz
Uncaught Error: Assertion failed: Trail visibility failure
Error: Assertion failed: Trail visibility failure
at window.assertions.assertFunction (http://127.0.0.1/continuous-testing/ct-snapshots/1683626850194/assert/js/assert.js:28:13)
at assertSlow (Instance.js:1886:6)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
id: Sparky Puppeteer
Snapshot from 5/9/2023, 4:07:30 AM

----------------------------------

center-and-variability : fuzz : unbuilt : assertSlow
http://127.0.0.1/continuous-testing/ct-snapshots/1683626850194/center-and-variability/center-and-variability_en.html?continuousTest=%7B%22test%22%3A%5B%22center-and-variability%22%2C%22fuzz%22%2C%22unbuilt%22%2C%22assertSlow%22%5D%2C%22snapshotName%22%3A%22snapshot-1683626850194%22%2C%22timestamp%22%3A1683636076375%7D&brand=phet&eall&fuzz
Query: brand=phet&eall&fuzz
Uncaught Error: Assertion failed: Trail visibility failure
Error: Assertion failed: Trail visibility failure
at window.assertions.assertFunction (http://127.0.0.1/continuous-testing/ct-snapshots/1683626850194/assert/js/assert.js:28:13)
at assertSlow (Instance.js:1886:6)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
id: Sparky Puppeteer
Snapshot from 5/9/2023, 4:07:30 AM

----------------------------------

center-and-variability : fuzz : unbuilt : assertSlow
http://127.0.0.1/continuous-testing/ct-snapshots/1683626850194/center-and-variability/center-and-variability_en.html?continuousTest=%7B%22test%22%3A%5B%22center-and-variability%22%2C%22fuzz%22%2C%22unbuilt%22%2C%22assertSlow%22%5D%2C%22snapshotName%22%3A%22snapshot-1683626850194%22%2C%22timestamp%22%3A1683638170985%7D&brand=phet&eall&fuzz
Query: brand=phet&eall&fuzz
Uncaught Error: Assertion failed
Error: Assertion failed
at window.assertions.assertFunction (http://127.0.0.1/continuous-testing/ct-snapshots/1683626850194/assert/js/assert.js:28:13)
at assert (Instance.js:1395:16)
at listener (TinyEmitter.ts:119:8)
at emit (Node.ts:882:30)
at insertChild (Node.ts:1089:15)
at setChildren (Node.ts:1116:9)
at (HomeScreenButton.ts:149:28)
at listener (TinyEmitter.ts:119:8)
at emit (ReadOnlyProperty.ts:315:22)
at _notifyListeners (ReadOnlyProperty.ts:266:13)
id: Sparky Puppeteer
Snapshot from 5/9/2023, 4:07:30 AM

----------------------------------

center-and-variability : fuzz : unbuilt : assertSlow
http://127.0.0.1/continuous-testing/ct-snapshots/1683626850194/center-and-variability/center-and-variability_en.html?continuousTest=%7B%22test%22%3A%5B%22center-and-variability%22%2C%22fuzz%22%2C%22unbuilt%22%2C%22assertSlow%22%5D%2C%22snapshotName%22%3A%22snapshot-1683626850194%22%2C%22timestamp%22%3A1683640338998%7D&brand=phet&eall&fuzz
Query: brand=phet&eall&fuzz
Uncaught Error: Assertion failed: Trail visibility failure
Error: Assertion failed: Trail visibility failure
at window.assertions.assertFunction (http://127.0.0.1/continuous-testing/ct-snapshots/1683626850194/assert/js/assert.js:28:13)
at assertSlow (Instance.js:1886:6)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
at auditVisibility (Instance.js:1896:22)
id: Sparky Puppeteer
Snapshot from 5/9/2023, 4:07:30 AM

----------------------------------

center-and-variability : phet-io-fuzz : unbuilt
http://127.0.0.1/continuous-testing/ct-snapshots/1683626850194/center-and-variability/center-and-variability_en.html?continuousTest=%7B%22test%22%3A%5B%22center-and-variability%22%2C%22phet-io-fuzz%22%2C%22unbuilt%22%5D%2C%22snapshotName%22%3A%22snapshot-1683626850194%22%2C%22timestamp%22%3A1683632090291%7D&ea&brand=phet-io&phetioStandalone&fuzz
Query: ea&brand=phet-io&phetioStandalone&fuzz
Uncaught Error: Assertion failed: stateObject is not valid for x
Error: Assertion failed: stateObject is not valid for x
at window.assertions.assertFunction (http://127.0.0.1/continuous-testing/ct-snapshots/1683626850194/assert/js/assert.js:28:13)
at assert (StateSchema.ts:147:34)
at Array.forEach
at forEach (StateSchema.ts:140:28)
at checkStateObjectValid (IOType.ts:403:42)
at isStateObjectValid (IOType.ts:436:9)
at validateStateObject (IOType.ts:281:23)
at toStateObject (ReadOnlyProperty.ts:306:34)
at getData (PhetioObject.ts:467:45)
at phetioStartEvent (ReadOnlyProperty.ts:301:66)
id: Sparky Puppeteer
Snapshot from 5/9/2023, 4:07:30 AM

----------------------------------

center-and-variability : phet-io-fuzz : unbuilt
http://127.0.0.1/continuous-testing/ct-snapshots/1683626850194/center-and-variability/center-and-variability_en.html?continuousTest=%7B%22test%22%3A%5B%22center-and-variability%22%2C%22phet-io-fuzz%22%2C%22unbuilt%22%5D%2C%22snapshotName%22%3A%22snapshot-1683626850194%22%2C%22timestamp%22%3A1683636396386%7D&ea&brand=phet-io&phetioStandalone&fuzz
Query: ea&brand=phet-io&phetioStandalone&fuzz
Uncaught Error: Assertion failed: stateObject is not valid for x
Error: Assertion failed: stateObject is not valid for x
at window.assertions.assertFunction (http://127.0.0.1/continuous-testing/ct-snapshots/1683626850194/assert/js/assert.js:28:13)
at assert (StateSchema.ts:147:34)
at Array.forEach
at forEach (StateSchema.ts:140:28)
at checkStateObjectValid (IOType.ts:403:42)
at isStateObjectValid (IOType.ts:436:9)
at validateStateObject (IOType.ts:281:23)
at toStateObject (ReadOnlyProperty.ts:306:34)
at getData (PhetioObject.ts:467:45)
at phetioStartEvent (ReadOnlyProperty.ts:301:66)
id: Sparky Puppeteer
Snapshot from 5/9/2023, 4:07:30 AM

----------------------------------

center-and-variability : phet-io-fuzz : unbuilt
http://127.0.0.1/continuous-testing/ct-snapshots/1683626850194/center-and-variability/center-and-variability_en.html?continuousTest=%7B%22test%22%3A%5B%22center-and-variability%22%2C%22phet-io-fuzz%22%2C%22unbuilt%22%5D%2C%22snapshotName%22%3A%22snapshot-1683626850194%22%2C%22timestamp%22%3A1683640695170%7D&ea&brand=phet-io&phetioStandalone&fuzz
Query: ea&brand=phet-io&phetioStandalone&fuzz
Uncaught Error: Assertion failed: stateObject is not valid for x
Error: Assertion failed: stateObject is not valid for x
at window.assertions.assertFunction (http://127.0.0.1/continuous-testing/ct-snapshots/1683626850194/assert/js/assert.js:28:13)
at assert (StateSchema.ts:147:34)
at Array.forEach
at forEach (StateSchema.ts:140:28)
at checkStateObjectValid (IOType.ts:403:42)
at isStateObjectValid (IOType.ts:436:9)
at validateStateObject (IOType.ts:281:23)
at toStateObject (ReadOnlyProperty.ts:306:34)
at getData (PhetioObject.ts:467:45)
at phetioStartEvent (ReadOnlyProperty.ts:301:66)
id: Sparky Puppeteer
Snapshot from 5/9/2023, 4:07:30 AM
```

**Comment by samreid** [comment]
@jonathanolson or @zepumph any ideas or recommendations about what is causing this, how to trace it or how to address it?

**Comment by zepumph** [comment]
I can get this paused in a debugger using the reported URL:

https://sparky.colorado.edu/continuous-testing/ct-snapshots/1684196480344/center-and-variability/center-and-variability_en.html?continuousTest=%7B%22test%22%3A%5B%22center-and-variability%22%2C%22fuzz%22%2C%22unbuilt%22%2C%22assertSlow%22%5D%2C%22snapshotName%22%3A%22snapshot-1684196480344%22%2C%22timestamp%22%3A1684198496256%7D&brand=phet&eall&fuzz

Probably the above link will work for ~24hours. Is that helpful? In the debugger I see that the trail looks like this. Perhaps that will help see the strangeness:



![image](https://github.com/phetsims/center-and-variability/assets/6856943/f4d02b71-42c5-49f1-aefe-dfde54621195)

**Comment by samreid** [comment]
Thanks, I was able to reproduce the bug with that link. In the dev tools, I wrote `this.node.parent.tandem` and it said:

> "centerAndVariability.optionalTandem.plotNode1.toggleNode.elements.rangeNode.rangeNode"

**Comment by matthew-blackman** [comment]
@samreid and I plan to add some debugging / loggin code to help get a trail from CT the next time this error pops up.

**Comment by matthew-blackman** [comment]
@samreid and I were able to track this down to the RangeNode hasNonZeroRange variable type, which was being set as 'undefined' if the leftmostDot or rightmostDot did not exist yet. This was causing undefined to be passed into the visible properties of rangeRectangle, rangeBar and rangeTextReadout which was causing an assertion error when assertSlow checked the trail visibility.

To fix this, we added type declarations to leftmostDot and rightmostDot, allowing them to be undefined. When determining hasNonZeroRange, we cast these values as booleans to ensure that boolean values will be used in setting node visibility.

We feel that this issue has been sufficiently tracked down and fixed. Closing.
